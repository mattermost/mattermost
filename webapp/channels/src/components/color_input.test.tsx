// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {renderWithContext, userEvent, fireEvent, screen} from 'tests/react_testing_utils';

import ColorInput from './color_input';

describe('components/ColorInput', () => {
    const baseProps = {
        id: 'sidebarBg',
        onChange: jest.fn(),
        value: '#ffffff',
    };

    test('should hide color picker when first rendered', () => {
        const {getByTestId} = renderWithContext(<ColorInput {...baseProps}/>);

        const inputElement = getByTestId('color-inputColorValue');

        expect(inputElement).toBeInTheDocument();
        expect(inputElement).toBeVisible();
        expect(screen.queryByTestId('color-popover')).not.toBeInTheDocument();
    });

    test('should toggle color picker when color picker button is clicked', () => {
        const {getByTestId} = renderWithContext(<ColorInput {...baseProps}/>);

        const colorPickerToggleButton = getByTestId('color-togglerButton');

        fireEvent.click(colorPickerToggleButton);

        const colorPopover = getByTestId('color-popover');

        expect(colorPopover).toBeInTheDocument();
        expect(colorPopover).toBeVisible();
        expect(document.activeElement).toBe(getByTestId('color-inputColorValue'));

        fireEvent.click(colorPickerToggleButton);

        expect(screen.queryByTestId('color-popover')).not.toBeInTheDocument();
    });

    test('should change color when the color picker is clicked', () => {
        const {getByTestId} = renderWithContext(<ColorInput {...baseProps}/>);

        fireEvent.click(getByTestId('color-togglerButton'));

        const colorPopover = getByTestId('color-popover');

        userEvent.click(colorPopover);

        expect(colorPopover).toBeInTheDocument();
        expect(baseProps.onChange).toHaveBeenCalledTimes(1);
    });

    test('should keep what the user types in the textbox until blur', () => {
        const {getByTestId} = renderWithContext(<ColorInput {...baseProps}/>);

        const inputElement = getByTestId('color-inputColorValue');

        inputElement.focus();

        expect(document.activeElement).toBe(inputElement);

        fireEvent.change(inputElement, {target: {value: '#abc'}});

        expect(inputElement).toHaveValue('#abc');
        expect(baseProps.onChange).toHaveBeenLastCalledWith('#aabbcc');

        // The RGB here is the equivalent of '#abc'.
        expect(getByTestId('color-icon').style.backgroundColor).toBe('rgb(170, 187, 204)');

        inputElement.blur();

        expect(document.activeElement).not.toBe(inputElement);
        expect(inputElement).toHaveValue('#aabbcc');
        expect(baseProps.onChange).toHaveBeenLastCalledWith('#aabbcc');

        // The RGB value passed in the assertion is the equivalent of '#aabbcc'.
        expect(getByTestId('color-icon').style.backgroundColor).toBe('rgb(170, 187, 204)');
    });
});
