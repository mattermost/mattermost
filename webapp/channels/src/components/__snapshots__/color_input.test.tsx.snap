// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/ColorInput should match snapshot, click on picker 1`] = `
<div
  className="color-input input-group"
>
  <input
    className="form-control"
    data-testid="color-inputColorValue"
    id="sidebarBg-inputColorValue"
    maxLength={7}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="#ffffff"
  />
  <span
    className="input-group-addon color-pad"
    id="sidebarBg-squareColorIcon"
    onClick={[Function]}
  >
    <i
      className="color-icon"
      id="sidebarBg-squareColorIconValue"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    />
  </span>
  <div
    className="color-popover"
    id="sidebarBg-ChromePickerModal"
  >
    <ColorPicker
      color="#ffffff"
      disableAlpha={true}
      onChange={[Function]}
      styles={Object {}}
      width={225}
    />
  </div>
</div>
`;

exports[`components/ColorInput should match snapshot, init 1`] = `
<div
  className="color-input input-group"
>
  <input
    className="form-control"
    data-testid="color-inputColorValue"
    id="sidebarBg-inputColorValue"
    maxLength={7}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="#ffffff"
  />
  <span
    className="input-group-addon color-pad"
    id="sidebarBg-squareColorIcon"
    onClick={[Function]}
  >
    <i
      className="color-icon"
      id="sidebarBg-squareColorIconValue"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    />
  </span>
</div>
`;

exports[`components/ColorInput should match snapshot, opened 1`] = `
<div
  className="color-input input-group"
>
  <input
    className="form-control"
    data-testid="color-inputColorValue"
    id="sidebarBg-inputColorValue"
    maxLength={7}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="#ffffff"
  />
  <span
    className="input-group-addon color-pad"
    id="sidebarBg-squareColorIcon"
    onClick={[Function]}
  >
    <i
      className="color-icon"
      id="sidebarBg-squareColorIconValue"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    />
  </span>
  <div
    className="color-popover"
    id="sidebarBg-ChromePickerModal"
  >
    <ColorPicker
      color="#ffffff"
      disableAlpha={true}
      onChange={[Function]}
      styles={Object {}}
      width={225}
    />
  </div>
</div>
`;

exports[`components/ColorInput should match snapshot, toggle picker 1`] = `
<div
  className="color-input input-group"
>
  <input
    className="form-control"
    data-testid="color-inputColorValue"
    id="sidebarBg-inputColorValue"
    maxLength={7}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="#ffffff"
  />
  <span
    className="input-group-addon color-pad"
    id="sidebarBg-squareColorIcon"
    onClick={[Function]}
  >
    <i
      className="color-icon"
      id="sidebarBg-squareColorIconValue"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    />
  </span>
</div>
`;
